mood_bot.py

```diff
@@ -0,0 +1,726 @@
+    else:
+        date_data = diary['date']
+        number_of_entries = len(date_data)
+        if number_of_entries == 0:
+            await context.send(diary_empty_message)
+        elif number_of_entries == 1:
+            await context.send(f'There is {str(number_of_entries)} entry in the mood diary.')
+            await context.send(f'This entry was on {str(date_data[0])}.')
+        else:
+            await context.send(f'There are {str(number_of_entries)} entries in the mood diary.')
+            await context.send(f'The first entry was on {str(date_data[0])} '
+                               f'and the last entry was on {str(date_data[number_of_entries - 1])}')
+    diary.close()
+
+
+async def diary_integer_error(context, start, end):
+    await context.send(integer_message_for_diary)
+    raise me.InputMoodError(integer_message_for_diary, [start, end])
+
+
+async def diary_range_error(context, start, end, number_of_entries):
+    out_of_diary_range_message_full = out_of_diary_range_message_partial + str(number_of_entries) + '.'
+    await context.send(out_of_diary_range_message_full)
+    raise me.InputMoodError(out_of_diary_range_message_full, [start, end])
+
+
+@mood_bot.command(name='print_diary', help="Prints out entries from the mood diary. "
+                                           "User specifies the start and end entries "
+                                           "or specifies 'all' for all entries.",
+                  aliases=['diary_print', 'printdiary', 'diaryprint'])
+async def print_diary(context, start='all', end='0'):
+    """Prints out the mood diary as specified by the user.
+    User can either write 'all' for all entries, or can specify start and end entries."""
+    author = context.author
+    diary_filename = str(author) + '_mood_diary'
+    diary = shelve.open(diary_filename)
+    if 'date' not in diary:
+        await context.send(diary_empty_message)
+    else:
+        diary.close()  # so that we don't try opening the diary file twice, concurrently
+        await mood_reset(context, False)  # makes sure that all the data is the same length
+        diary = shelve.open(diary_filename)
+        [date_data, type_data, rating_data, why_data] = mass_diary_data_retrieval(diary)
+        number_of_entries = len(date_data)
+        if start.lower() == 'all':
+            start = 1
+            end = number_of_entries
+        try:
+            start = int(start)
+            end = int(end)
+        except ValueError:
+            diary.close()
+            await diary_integer_error(context, start, end)
+        if start < 1 or end > number_of_entries:
+            await diary_range_error(context, start, end, number_of_entries)
+        else:
+            start = start - 1
+            for j in range(start, end):
+                entry = [str(date_data[j]), str(type_data[j]), str(rating_data[j]), str(why_data[j])]
+                why_string = entry[3]
+                why_string = str(why_string)

```

Comment:
look up escape characters maybe? maybe that works? @PhilippRisius ?

Reply:
So, I was trying to '.join' the 'why' statement in this function which led to all these extra punctuation for some reason. But, when I use the '.join' function in the why function, it works properly and converts the list to a string.

Reply:
So, I no longer have all these replace functions.